Mathlib/RingTheory/TwoSidedIdeal/Operations.lean

```diff
@@ -0,0 +1,207 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.RingTheory.TwoSidedIdeal.Lattice
+import Mathlib.RingTheory.TwoSidedIdeal.BigOperators
+import Mathlib.RingTheory.Congruence.Opposite
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.Data.Fintype.BigOperators
+import Mathlib.RingTheory.Ideal.Basic
+import Mathlib.Order.GaloisConnection
+
+
+/-!
+# Operations on Two-Sided-Ideals
+
+This file defines operations on two-sided-ideals of a ring `R`.
+
+## Main definitions and results
+- `TwoSidedIdeal.span`: the span of `s ⊆ R` is the smallest two-sided-ideal containing the set.
+- `TwoSidedIdeal.mem_span_iff_exists`: an element `x` is in the span of `s` if and only if it can be
+  written as a finite sum of the form `∑ i, xL i * y i * xR i` where `xL i ∈ s`, `y i ∈ R`, and
+  `xR i ∈ s`.
+
+- `TwoSidedIdeal.comap`: pullback of a two-sided-ideal; defined as the preimage of a
+  two-sided-ideal.
+- `TwoSidedIdeal.map`: pushforward of a two-sided-ideal; defined as the span of the image of a
+  two-sided-ideal.
+- `TwoSidedIdeal.ker`: the kernel of a ring homomorphism as a two-sided-ideal.
+-/
+
+namespace TwoSidedIdeal
+
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R]
+
+/--
+The smallest two-sided-ideal containing a set.
+-/
+abbrev span (s : Set R) : TwoSidedIdeal R :=
+  { ringCon := ringConGen (fun a b ↦ a - b ∈ s) }
+
+lemma subset_span {s : Set R} : s ⊆ (span s : Set R) := by
+  intro x hx
+  rw [SetLike.mem_coe, mem_iff]
+  exact RingConGen.Rel.of _ _ (by simpa using hx)
+
+lemma mem_span_iff {s : Set R} {x} :
+    x ∈ span s ↔ ∀ (I : TwoSidedIdeal R), s ⊆ I → x ∈ I := by
+  refine ⟨?_, fun h => h _ subset_span⟩
+  delta span
+  rw [RingCon.ringConGen_eq]
+  intro h I hI
+  refine sInf_le (α := RingCon R) ?_ h
+  intro x y hxy
+  specialize hI hxy
+  rwa [SetLike.mem_coe, ← rel_iff] at hI
+
+lemma span_mono {s t : Set R} (h : s ⊆ t) : span s ≤ span t := by
+  intro x hx
+  rw [mem_span_iff] at hx ⊢
+  exact fun I hI => hx I $ h.trans hI
+
+end NonUnitalNonAssocRing
+
+section NonAssocRing
+
+variable {R S : Type*} [NonAssocRing R] [NonAssocRing S]
```

Comment:
Why exactly do you require the rings in this section to be unital? All of these concepts should work for non-unital rings too.

Reply:
These construction needs the concept of `RingHomClass`, which requires a `1`.

Reply:
Okay, but they shouldn't. If they do, it's a deficiency with `RingCon` that needs to be fixed first.

Reply:
This has been relaxed